substrate/primitives/staking/src/delegation.rs

```diff
@@ -0,0 +1,108 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::StakingInterface;
+use sp_runtime::{DispatchError, DispatchResult};
+
+/// A generic representation of a delegation based staking apis that other runtime pallets can use.
+///
+/// Compared to StakingInterface that allows an account to be a direct nominator,
+/// `DelegateStakingInterface` allows an account (called delegator) to delegate its stake to another
+/// account (delegatee). In delegation based staking, the funds are locked in the delegator's
+/// account and gives the delegatee the right to use the funds for staking as if it is a direct
+/// nominator.
+pub trait DelegatedStakeInterface: StakingInterface {
+	/// Set intention to accept delegations.
+	///
+	/// Takes payee as input which determines where reward should be paid out.
+	///
+	/// Once executed, the account is registered as a `Delegatee`.
+	fn accept_delegations(delegatee: &Self::AccountId, payee: &Self::AccountId) -> DispatchResult;
+
+	/// Stop accepting new delegations on this account.
+	fn block_delegations(delegatee: &Self::AccountId) -> DispatchResult;
+
+	/// Remove oneself as Delegatee.
+	///
+	/// This will only succeed if all delegations to this delegatee are removed.
+	fn kill_delegatee(delegatee: &Self::AccountId) -> DispatchResult;
+
+	/// Delegate some funds to a Delegatee
+	///
+	/// Delegated funds are locked in delegator's account and added to Delegatee's stakeable
+	/// balance.
+	fn delegate(
+		delegator: &Self::AccountId,
+		delegatee: &Self::AccountId,
+		value: Self::Balance,
+	) -> DispatchResult;
+
+	/// Update bond whenever there is a new delegate funds that are not staked.
+	fn update_bond(delegatee: &Self::AccountId) -> DispatchResult;
+
+	/// Unbond some funds from a delegatee.
+	///
+	/// Similar to [`StakingInterface::unbond`].
+	fn unbond(delegatee: &Self::AccountId, value: Self::Balance) -> DispatchResult;
```

Comment:
Would this interface need to define another `fn unbound` or could the logic be abstracted so that it relies on the `StakingInterface::unbound`? In that case, the account ID of the delegator/stash is be the only thing that implicitly has the information of whether the account is a delegatee or a direct nomination stash and that should be handled seamsly, if possible.

I think the overall comment here is that having a trait (`DelegatedStakeInterface`) that depends on another (`StakingInterface`) with the same methods seems a bit wrong to me.

Reply:
Yes I think we could remove this dependency. I started with this thinking DelegatedStakeInterface will be used by NP but now I think we can have a StakingDelegationInterface to add or remove delegation, and then use the normal StakeInterface for NP.